plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,286 @@
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = clusterDetails.map(c => {
+      const cAuth: ClusterDetails = Object.assign(c, {
+        serviceAccountToken: krc[c.name],
+      });
+      return cAuth;
+    });
+
+    return clusterDetailsAuth;
+  }
+
+  private getClusterURI(details: ClusterDetails): string {
+    const client = this.getKubeConfig(details);
+    return client.getCurrentCluster()?.server || '';
+  }
+
+  private async makeRequestToCluster(
+    details: ClusterDetails,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const serverIP = this.getClusterURI(details);
+    if (!serverIP) {
+      return {
+        code: ERROR_INTERNAL_SERVER,
+        data: null,
+      };
+    }
+
+    const query = decodeURIComponent(req.params.encodedQuery) || '';
+    const uri = `${serverIP}/${query}`;
+
+    const contentType = req.header(HEADER_CONTENT_TYPE) || APPLICATION_JSON;
+
+    const res = await this.sendClusterRequest(
+      details,
+      uri,
+      req.method,
+      contentType,
+      req.body,
+    );
+
+    return res;
+  }
+
+  private async sendClusterRequest(
+    details: ClusterDetails,
+    uri: string,
+    method: string,
+    contentType: string,
+    body?: any,
+  ): Promise<KubernetesProxyResponse> {
+    const bearerToken = details.serviceAccountToken;
+    if (!bearerToken) {
+      return {
+        code: ERROR_BAD_REQUEST,
+        data: {
+          error: 'Invalid service account token',
+        },
+      };
+    }
+
+    const reqData: any = {
+      method,
+      headers: {
+        'Content-Type': contentType,
+        Authorization: `Bearer ${bearerToken}`,
+      },
+    };
+
+    if (!details.skipTLSVerify) {
+      if (details.caData) {
+        const ca = bufferFromFileOrString('', details.caData)?.toString() || '';
+        reqData.agent = new https.Agent({ ca });
+      } else {
+        this.logger.info('could not find CA certificate!');
+        return {
+          code: ERROR_INTERNAL_SERVER,
+          data: {
+            error: 'Invalid CA certificate configured within Backstage',
+          },
+        };
+      }
+    }
+
+    if (body && Object.keys(body).length > 0) {
+      reqData.body = JSON.stringify(body);
+    }
+
+    try {
+      const req = await fetch(uri, reqData);
+
+      let res;
+      if (contentType.includes(APPLICATION_JSON)) {
+        res = await req.json();
+      } else {
+        res = await req.text();
+      }
+
+      const proxyResponse: KubernetesProxyResponse = {
+        code: req.status,
+        data: res,
+        cluster: details.name,
+      };
+
+      return proxyResponse;
+    } catch (e: any) {
+      return {
+        code: ERROR_INTERNAL_SERVER,
```

Comment:
What happens if the upstream responds with a 204 and no payload, will that be translated into a 500 here, does .json() throw then? Maybe it's not a problem in practice, just checking. This is the type of thing that's easy to test with MSW too